contracts/Staking.sol

```diff
@@ -0,0 +1,177 @@
+    /// @param _finish the finish date of stake
+    /// @param _rewardsAmount the amount of reward tokens
+    /// @param _apy the annual percentage yield
+    function setRewards(
+        uint256 _start,
+        uint256 _finish,
+        uint256 _rewardsAmount,
+        uint256 _apy
+    ) external virtual override onlyOwner {
+        require(
+            _start < _finish && block.timestamp < _start,
+            "Incorrect dates"
+        );
+        require(_apy != 0 && _apy <= 10, "Incorrect apy");
+        require(
+            _rewardsAmount != 0 && _rewardsAmount / _apy == 50000 * DECIMALS18,
+            "Incorrect amount of tokens"
+        );
+
+        rewardToken.safeTransferFrom(msg.sender, address(this), _rewardsAmount);
+        startDate = _start;
+        endDate = _finish;
+        apy = _apy;
+        emit SetRewards(_start, _finish, _rewardsAmount, _apy);
+    }
+
+    /// @notice Transfer the amount of tokens from the user account and register staking for him
+    /// @dev Transfer the amount of tokens from the user account and register staking for him
+    /// @param _amount the amount of staked tokens
+    function stake(uint256 _amount) external virtual override {
+        require(_amount > 0, "ERROR_AMOUNT_IS_ZERO");
+        require(
+            totalSupply + _amount <= 5_000_000 * DECIMALS18,
+            "ERROR_MAX_TOTAL_SUPPLY"
+        );
+        require(
+            block.timestamp > stakeTime[msg.sender] + stakingInfo.cooldown,
+            "The cooldown has not been passed"
+        );
+        rewards[msg.sender] += calculateRewardTokens(
+            msg.sender,
+            staked[msg.sender]
+        );
+        stakeTime[msg.sender] = block.timestamp;
+        staked[msg.sender] += _amount;
+        totalSupply += _amount;
+        stakingToken.safeTransferFrom(msg.sender, address(this), _amount);
+        emit Stake(msg.sender, _amount);
+    }
+
+    /// @notice Transfer all staked tokens and rewards to the user account and update staking details for him
+    /// @dev Transfer all staked tokens and rewards to the user account and update staking details for him
+    function unstake() external virtual override {
+        uint256 amountToTransfer = staked[msg.sender];
+        require(amountToTransfer != 0, "There is no staked tokens");
+        rewards[msg.sender] += calculateRewardTokens(
+            msg.sender,
+            amountToTransfer
+        );
+        staked[msg.sender] = 0;
+        totalSupply -= amountToTransfer;
+        stakingToken.safeTransfer(msg.sender, amountToTransfer);
+        amountToTransfer = rewards[msg.sender];
+        if (
+            block.timestamp < stakeTime[msg.sender] + stakingInfo.stakingPeriod
+        ) {
+            amountToTransfer =
+                (amountToTransfer * (100 - stakingInfo.feePercentage)) /
+                100;
+        }
+        rewards[msg.sender] = 0;
+        rewardToken.safeTransfer(msg.sender, amountToTransfer);
```

Comment:
if balance of rewardToken is less the amountToTransfer, just transfer rest of balance